File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    vector<vector<int>> dp(triangle.size(),vector<int>(triangle.size(),0));
    vector<vector<int>> v;
    for(int i = 0;i<triangle.size();i++){
        v.push_back(triangle[i]);
    }
    dp[0][0] = v[0][0];
    for(int i = 0;i<triangle.size()-1;i++){
        for(int j =0;j<i+1;j++){
            dp[i+1][j] = max(dp[i+1][j], dp[i][j] + v[i+1][j]);
            dp[i+1][j+1]  = max(dp[i+1][j+1],dp[i][j] + v[i+1][j+1]);
        }
    }
    

    answer = *max_element(dp[dp.size()-1].begin(),dp[dp.size()-1].end());
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    var answer = 0;
    var leng = triangle.length;
    var dp = Array.from({length: leng}, () => new Array(leng).fill(0));
    
    dp[0][0] = triangle[0][0];
    

    for(let i = 1;i<triangle.length;i++){
        for(let j = 0;j<triangle[i].length;j++){
            if(j==0){
                dp[i][j] = dp[i-1][j] + triangle[i][j];
            }else{
            dp[i][j] = Math.max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j];
                                
            }
        }
    }
        for(var row of dp){
        for(var v of row){
            //console.log(v);
        }
    }
    for(let i = 0;i<leng;i++){
        answer = Math.max(answer, dp[leng-1][i]);
    }
    return answer;
}

//7
//3 8
//8 1 0
//2 7 4 4
//4 5 2 6 5

//7
//10 15
//18 16
